# __Canadian Federal Government Contract Spending Audit Case Study__

This case study analyzes data from the Government of Canada's Open Government Portal, with a specific focus on contract awards. The analysis utilizes data from the 2024-2025 fiscal year, representing the most recent complete dataset available. The case study is structured into three distinct parts, as outlined below:
1. Data acquisition and preparation which will deal with downloading the data and cleaning it for further analysis.
2.	Audit Objectives & Hypotheses section which deals with creating and testing different hypotheses regarding the contract award process.
3.	Data Analysis & Insights section where PowerBi will be used to identify and describe valuable audit or compliance-related insights.



## Section 1. Data Acquisition and preparation.

Data for the 2024-2025 fiscal year was downloaded from the Government of Canada's Open Government Portal, as a CSV file and loaded into a pandas’ data frame. Python will be used for this section.


In [40]:
# Importing Functions
import pandas as pd
import numpy as np



In [41]:
# Load the dataframe.
df_award_notice = pd.read_csv('2024-2025-awardNotice-avisAttribution.csv')

# Display the first 5 rows.
display(df_award_notice.head())
df_award_notice.info()
df_award_notice.describe()

,title-titre-eng,title-titre-fra,referenceNumber-numeroReference,amendmentNumber-numeroModification,solicitationNumber-numeroSollicitation,contractNumber-numeroContrat,publicationDate-datePublication,contractAwardDate-dateAttributionContrat,amendmentDate-dateModification,contractStartDate-contratDateDebut,...,contactInfoCity-contacterInfoVille-eng,contactInfoProvince-contacterInfoProvince-eng,contactInfoPostalcode,contactInfoCountry-contactInfoPays-eng,contactInfoAddressLine-contactInfoAdresseLigne-fra,contactInfoCity-contacterInfoVille-fra,contactInfoProvince-contacterInfoProvince-fra,contactInfoCountry-contactInfoPays-fra,awardDescription-descriptionAttribution-eng,awardDescription-descriptionAttribution-fra
0,"PRE-QUALIFICATION FOR ORM CONSULTING, ADVISORY...",PRÉQUALIFICATION POUR DES SERVICES DE CONSULTA...,MX-43089535758,0,PS23-AG-023,NaN,2024-04-10,2024-04-01,NaN,2024-04-01,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,FCC&rsquo;s intention through this RFP is to e...,"L&rsquo;intention de FAC, dans le cadre de la ..."
1,"PRE-QUALIFICATION FOR ORM CONSULTING, ADVISORY...",PRÉQUALIFICATION POUR DES SERVICES DE CONSULTA...,MX-43089535759,0,PS23-AG-023,NaN,2024-04-10,2024-04-01,NaN,2024-04-01,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,FCC&rsquo;s intention through this RFP is to e...,"L&rsquo;intention de FAC, dans le cadre de la ..."
2,"PRE-QUALIFICATION FOR ORM CONSULTING, ADVISORY...",PRÉQUALIFICATION POUR DES SERVICES DE CONSULTA...,MX-43089535760,0,PS23-AG-023,NaN,2024-04-10,2024-04-01,NaN,2024-04-01,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,FCC&rsquo;s intention through this RFP is to e...,"L&rsquo;intention de FAC, dans le cadre de la ..."
3,HUMAN CENTRED DESIGN,CONCEPTION CENTRÉE SUR LA PERSONNE,MX-443104980018,0,PS24-AG-001,NaN,2024-04-24,2024-04-24,NaN,2024-04-15,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,FCC has an in-house design thinking / human ce...,FAC dispose d&rsquo;une fonction interne de co...
4,RFP – PRE-QUALIFICATION FOR STRATEGIC TRANSFOR...,DDP – PRÉQUALIFICATION POUR DES SERVICES DE TR...,MX-443147876056,0,PS23-AG-018,NaN,2024-06-04,2024-02-05,NaN,2024-02-05,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"Effective July 31, 2023, FCC established a Tra...","Le 31 juillet 2023, FAC a cr&eacute;&eacute; u..."


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6572 entries, 0 to 6571
Data columns (total 76 columns):
 #   Column                                                                  Non-Null Count  Dtype  
---  ------                                                                  --------------  -----  
 0   title-titre-eng                                                         6572 non-null   object 
 1   title-titre-fra                                                         6572 non-null   object 
 2   referenceNumber-numeroReference                                         6572 non-null   object 
 3   amendmentNumber-numeroModification                                      6572 non-null   int64  
 4   solicitationNumber-numeroSollicitation                                  6572 non-null   object 
 5   contractNumber-numeroContrat                                            6439 non-null   object 
 6   publicationDate-datePublication                                         6572 non

,amendmentNumber-numeroModification,contractAmount-montantContrat,totalContractValue-valeurTotaleContrat
count,6572.000000,4.158000e+03,6.303000e+03
mean,1.249848,7.502075e+05,3.817224e+06
std,2.207324,1.146980e+07,1.457078e+08
min,0.000000,-3.372380e+06,0.000000e+00
25%,0.000000,0.000000e+00,0.000000e+00
50%,1.000000,0.000000e+00,1.031864e+05
75%,1.000000,0.000000e+00,3.982267e+05
max,26.000000,6.532125e+08,1.121937e+10



Initial analysis of the dataset reveals the following:
1. 	There is a total of 76 columns and 6571 rows. The columns have the names in both English and French with a dash in between.
2. 	Only 3 columns have a numeric datatype, the rest have an object datatype including the date columns
  



### Data cleanup

The column titles are edited to remove the French names, and the duplicated columns are deleted. This leaves a total of 49 unique columns.


In [42]:
# Remove french names from the columns
df_award_notice.columns = [col.split('-')[0] for col in df_award_notice.columns]

# Find and remove duplicated columns
df_award_notice = df_award_notice.loc[:,~df_award_notice.columns.duplicated()]
display(df_award_notice.head())

,title,referenceNumber,amendmentNumber,solicitationNumber,contractNumber,publicationDate,contractAwardDate,amendmentDate,contractStartDate,contractEndDate,...,contactInfoName,contactInfoEmail,contactInfoPhone,contactInfoFax,contactInfoAddressLine,contactInfoCity,contactInfoProvince,contactInfoPostalcode,contactInfoCountry,awardDescription
0,"PRE-QUALIFICATION FOR ORM CONSULTING, ADVISORY...",MX-43089535758,0,PS23-AG-023,NaN,2024-04-10,2024-04-01,NaN,2024-04-01,2027-03-31,...,Adam Gottselig,procurement@fcc.ca,000-000-0000,NaN,NaN,NaN,NaN,NaN,NaN,FCC&rsquo;s intention through this RFP is to e...
1,"PRE-QUALIFICATION FOR ORM CONSULTING, ADVISORY...",MX-43089535759,0,PS23-AG-023,NaN,2024-04-10,2024-04-01,NaN,2024-04-01,2027-03-31,...,Adam Gottselig,procurement@fcc.ca,000-000-0000,NaN,NaN,NaN,NaN,NaN,NaN,FCC&rsquo;s intention through this RFP is to e...
2,"PRE-QUALIFICATION FOR ORM CONSULTING, ADVISORY...",MX-43089535760,0,PS23-AG-023,NaN,2024-04-10,2024-04-01,NaN,2024-04-01,2027-03-31,...,Adam Gottselig,procurement@fcc.ca,000-000-0000,NaN,NaN,NaN,NaN,NaN,NaN,FCC&rsquo;s intention through this RFP is to e...
3,HUMAN CENTRED DESIGN,MX-443104980018,0,PS24-AG-001,NaN,2024-04-24,2024-04-24,NaN,2024-04-15,2027-04-14,...,Adam Gottselig,procurement@fcc.ca,000-000-0000,NaN,NaN,NaN,NaN,NaN,NaN,FCC has an in-house design thinking / human ce...
4,RFP – PRE-QUALIFICATION FOR STRATEGIC TRANSFOR...,MX-443147876056,0,PS23-AG-018,NaN,2024-06-04,2024-02-05,NaN,2024-02-05,2027-02-04,...,Adam Gottselig,procurement@fcc.ca,000-000-0000,NaN,NaN,NaN,NaN,NaN,NaN,"Effective July 31, 2023, FCC established a Tra..."


#### Date Columns

##### Several columns, including publicationDate, contractAwardDate, amendmentDate, contractStartDate, and contractEndDate, were initially of object type and are converted to datetime objects.


In [43]:
# 2. Convert date columns to datetime objects
date_cols = ['publicationDate', 'contractAwardDate', 'amendmentDate', 'contractStartDate', 'contractEndDate']

for col in date_cols:
    df_award_notice[col] = pd.to_datetime(df_award_notice[col], errors='coerce')

print("Number of NaT values introduced after converting to datetime:")
for col in date_cols:
    # Re-calculate missing_values since the dataframe changed
    missing_values_after_drop = df_award_notice.isnull().sum()
    nat_count = df_award_notice[col].isnull().sum() - missing_values_after_drop[col]
    print(f"- {col}: {nat_count} NaT values")

Number of NaT values introduced after converting to datetime:
- publicationDate: 0 NaT values
- contractAwardDate: 0 NaT values
- amendmentDate: 0 NaT values
- contractStartDate: 0 NaT values
- contractEndDate: 0 NaT values


#### Inconsistent Data

Inconsistencies were found and addressed in the following categorical columns: procurementCategory, supplierAddressCountry, contractingEntityAddressProvince, contractingEntityAddressCountry, and contactInfoCountry. These inconsistencies included variations in spelling, capitalization, extra spaces, and non-standard entries.


Analyse categorical columns

In [44]:
# Analyse categorical columns
categorical_cols = df_award_notice.select_dtypes(include='object').columns

print("Number of unique values for each categorical column:")
for col in categorical_cols:
    num_unique = df_award_notice[col].nunique()
    print(f"- {col}: {num_unique} unique values")
    if num_unique < 50:
        print(f"  Unique values for {col}:")
        display(df_award_notice[col].unique())

Number of unique values for each categorical column:
- title: 4810 unique values
- referenceNumber: 6572 unique values
- solicitationNumber: 4768 unique values
- contractNumber: 5515 unique values
- contractCurrency: 5 unique values
  Unique values for contractCurrency:


array(['CAD', nan, 'USD', 'AUD', 'GBP', 'EUR'], dtype=object)

- awardStatus: 3 unique values
  Unique values for awardStatus:


array(['Active', 'Cancelled', 'Expired'], dtype=object)

- gsin: 103 unique values
- gsinDescription: 103 unique values
- unspsc: 1972 unique values
- unspscDescription: 1972 unique values
- procurementCategory: 6 unique values
  Unique values for procurementCategory:


array(['*SRV', '*GD', '*SRV\n*GD', '*SRVTGD', '*GD\n*SRV', '*CNST'],
      dtype=object)

- noticeType: 8 unique values
  Unique values for noticeType:


array([nan, 'Request for Proposal', 'RFP against Supply Arrangement',
       'Request for Standing Offer', 'Advance Contract Award Notice',
       'Not Applicable', 'Invitation to Qualify', 'Directed Contract',
       'Request for Supply Arrangement'], dtype=object)

- procurementMethod: 5 unique values
  Unique values for procurementMethod:


array(['Competitive - Open bidding', 'Non-competitive',
       'Competitive - Traditional', 'Competitive - Selective tendering',
       'Advance contract award notice', nan], dtype=object)

- selectionCriteria: 10 unique values
  Unique values for selectionCriteria:


array([nan, 'Lowest Price',
       'Highest Combined Rating of Technical Merit and Price',
       'Highest Technical Merit within a Stipulated Maximum Budget',
       'Variations or combinations of the above methods',
       'Not applicable', 'Lowest Cost-per-Point',
       'Government Objectives Representing Best Interests/Value to Govt',
       'Only One Response to Bid Solicitation', 'Low Dollar Value',
       'Subsequent/Follow-on Contracts'], dtype=object)

- limitedTenderingReason: 10 unique values
  Unique values for limitedTenderingReason:


array([nan, 'Exclusive Rights', 'Additional Deliveries',
       'Goods Purchased on a Commodity Market',
       'No response to bid solicitation', 'Extreme Urgency',
       'Purchases Made Under Exceptionally Advantageous Conditions',
       'Work on a leased building or related property performed only by the lessor (for contracts that are only covered by the CFTA)',
       'Prototype Purchase',
       'Consulting Services Regarding Matters of a Confidential Nature (for contracts not covered by CETA, CFTA, WTO-GPA, CPTPP, CKFTA and CUFTA)',
       'Subscriptions to newspapers, magazines, or other periodicals (for contracts that are only covered by the CFTA)'],
      dtype=object)

- tradeAgreements: 3010 unique values
- regionsOfDelivery: 655 unique values
- supplierLegalName: 4400 unique values
- supplierAddressLine: 4483 unique values
- supplierAddressCity: 1233 unique values
- supplierAddressProvince: 240 unique values
- supplierAddressPostalCode: 3713 unique values
- supplierAddressCountry: 75 unique values
- contractingEntityName: 249 unique values
- contractingEntityAddressLine: 468 unique values
- contractingEntityAddressCity: 80 unique values
- contractingEntityAddressProvince: 39 unique values
  Unique values for contractingEntityAddressProvince:


array(['Saskatchewan', 'Nova Scotia', 'Quebec', 'Ontario', 'QC', 'ON',
       'SK', 'On', 'Terre Neuve et Labrador', 'Manitoba', 'Alberta', 'NL',
       'New Brunswick', 'Québec', 'MB', 'Ontario ', 'NS', 'NB',
       'Quebec  ', 'Quebec ', 'BC', 'Qc', 'British Columbia', 'On ',
       'B.C.', 'Sasktchewan', 'ONTARIO', 'Ontario  ', nan, 'AB',
       'Ontarion', 'SK ', 'ON ', 'ON  ', 'Qatar', 'QUEBEC',
       'National Capital Region (NCR)', 'OB', 'Prince Edward Island',
       'Ottawa'], dtype=object)

- contractingEntityAddressPostalCode: 250 unique values
- contractingEntityAddressCountry: 9 unique values
  Unique values for contractingEntityAddressCountry:


array(['Canada', 'Canada ', 'CA', 'CANADA', 'canada', nan, 'CAN', 'Canda',
       ' Canada', 'Qatar'], dtype=object)

- endUserEntitiesName: 122 unique values
- endUserEntitiesAddress: 111 unique values
- contactInfoName: 1264 unique values
- contactInfoEmail: 1185 unique values
- contactInfoPhone: 624 unique values
- contactInfoFax: 40 unique values
  Unique values for contactInfoFax:


array([nan, '613-965-7615', '(855) 983-1808', '(866) 246-6893',
       'Not applicable', '1-866-246-6893', '000000', '1-855-983-1808',
       'N/A ', 'N/a', 'Not Applicable', '1-877-558-2349', '306-780-5232',
       '902-426-7136', '780-454-4523', '306-780-5018', '(819) 939-5042',
       '780-454-4527', '(416) 601-4739', '866-246-6893', 'not applicable',
       '604-870-2444', '(506) 851-6327', 'S/O', 'Sans objet',
       '1-866-246-6863', '613-960-6026', '613-123-4567', '3435769329',
       '613-866-5465', '613 960-6026', '6133015643', '00000000', '613-',
       '11111111', '(613) 960-6007', '1-800-613-1303', '613-555-5555',
       '5.14913e+009', '0', '6.13614e+009'], dtype=object)

- contactInfoAddressLine: 457 unique values
- contactInfoCity: 83 unique values
- contactInfoProvince: 56 unique values
- contactInfoPostalcode: 250 unique values
- contactInfoCountry: 11 unique values
  Unique values for contactInfoCountry:


array([nan, 'Canada ', 'Canada', 'canada', 'CA', 'Ca', 'United States',
       'CANADA', ' ', 'CAN', ' Canada', 'Canadaq'], dtype=object)

- awardDescription: 4060 unique values


#### Missing Values

The columns with missing values were analysed and changes were made as follows:
1. gsin ( Goods and Services Identification Number) column has 96% of the data missing and was deleted as this is a legacy code no longer used according to the official documentation. This is the same for the gsinDescription column as well. 
2.	A total of 7 columns containing contact info for tendering organisations or individuals had more than 70% of the data missing and these were deleted given the context. The columns are contactInfoAddressLine, contactInfoPostalcode, contactInfoProvince, contactInfoCity, contactInfoCountry, contactInfoPhone and contactInfoFax. Similarly named columns with supplier contact information had insignificant missing values indicating that these 7 columns are most likely duplicates.
3.	Some entries for contractAmount and totalContractValue had negative values and given the context of contract awards, these were most likely errors and were set to the absolute figure. Missing values in these 2 columns were imputed to 0.
4.	Non-datetime, non-numeric columns with missing values between 10% and 50%, missing values were imputed with the string 'Unknown'. For categorical columns in this range, 'Unknown' was added as a category before imputation.


In [45]:
# Show percentage of missing values per column
missing_values = df_award_notice.isnull().sum()
missing_percentage = (missing_values / len(df_award_notice)) * 100

missing_info = pd.DataFrame({'Missing Count': missing_values, 'Missing Percentage (%)': missing_percentage})
missing_info = missing_info[missing_info['Missing Count'] > 0].sort_values(by='Missing Count', ascending=False)

print("Missing values per column:")
display(missing_info)

Missing values per column:


,Missing Count,Missing Percentage (%)
contactInfoFax,6427,97.793670
gsin,6303,95.906878
gsinDescription,6303,95.906878
limitedTenderingReason,5716,86.975046
contactInfoAddressLine,4826,73.432745
contactInfoPostalcode,4823,73.387097
contactInfoProvince,4817,73.295800
contactInfoCity,4814,73.250152
contactInfoCountry,4770,72.580645
contractCurrency,4518,68.746196


In [47]:

#  Drop columns with very high percentage of missing values 
cols_to_drop = [
    'gsin', 'gsinDescription', 'contactInfoAddressLine', 'contactInfoPostalcode',
    'contactInfoProvince', 'contactInfoCity', 'contactInfoCountry', 'contactInfoPhone',
    'contactInfoFax'
]
cols_to_drop_existing = [col for col in cols_to_drop if col in df_award_notice.columns]

if cols_to_drop_existing:
    df_award_notice = df_award_notice.drop(columns=cols_to_drop_existing)
    print(f"Columns dropped: {cols_to_drop_existing}")
else:
    print("No specified columns with high missing values to drop.")

print(f"Number of columns after dropping: {df_award_notice.shape[1]}")

# Convert date columns to datetime objects (already done in previous step, but included for completeness)
date_cols = ['publicationDate', 'contractAwardDate', 'amendmentDate', 'contractStartDate', 'contractEndDate']
for col in date_cols:
    if col in df_award_notice.columns: # Check if column exists after dropping
        df_award_notice[col] = pd.to_datetime(df_award_notice[col], errors='coerce')


#  Address inconsistencies in categorical columns
# Standardize country names
country_cols = ['supplierAddressCountry', 'contractingEntityAddressCountry', 'contactInfoCountry']
for col in country_cols:
    if col in df_award_notice.columns: # Check if column exists after dropping
        df_award_notice[col] = df_award_notice[col].astype(str).str.strip().str.upper().replace({
            'CA': 'CANADA', 'CANDA': 'CANADA', 'CAN': 'CANADA', '': None, 'NAN': None
        })


# Standardize province names
province_cols = ['contractingEntityAddressProvince']
for col in province_cols:
     if col in df_award_notice.columns: # Check if column exists after dropping
        df_award_notice[col] = df_award_notice[col].astype(str).str.strip().replace({
            'QC': 'Quebec', 'Québec': 'Quebec', 'ON': 'Ontario', 'On': 'Ontario',
            'National Capital Region (NCR)': 'Ontario', 'Ottawa': 'Ontario',
            'Ontarion': 'Ontario', 'nan': None
        })


# Clean procurement category variations
if 'procurementCategory' in df_award_notice.columns: # Check if column exists after dropping
    df_award_notice['procurementCategory'] = df_award_notice['procurementCategory'].astype(str).replace({
        '\n': '', 'SRVTGD': 'SRV*GD'
    }, regex=True).str.replace(r'\*GD\*SRV', '*SRV*GD').replace('nan', None)


#  Handle missing values 
# Address negative 'contractAmount' and 'totalContractValue' by setting them to their absolute value
if 'contractAmount' in df_award_notice.columns:
    df_award_notice['contractAmount'] = df_award_notice['contractAmount'].abs()
if 'totalContractValue' in df_award_notice.columns:
    df_award_notice['totalContractValue'] = df_award_notice['totalContractValue'].abs()

# Impute missing 'contractAmount' and 'totalContractValue' with 0
if 'contractAmount' in df_award_notice.columns:
    df_award_notice['contractAmount'] = df_award_notice['contractAmount'].fillna(0)
if 'totalContractValue' in df_award_notice.columns:
    df_award_notice['totalContractValue'] = df_award_notice['totalContractValue'].fillna(0)

# Impute missing values in non-datetime, non-numeric columns (with missing values between 10% and 50%) with 'Unknown'
# Identify non-datetime, non-numeric columns
object_cols = df_award_notice.select_dtypes(include='object').columns
datetime_cols_existing = [col for col in date_cols if col in df_award_notice.columns]
numerical_cols_existing = [col for col in ['amendmentNumber', 'contractAmount', 'totalContractValue'] if col in df_award_notice.columns]

cols_to_impute_unknown = [col for col in object_cols if col not in datetime_cols_existing and col not in numerical_cols_existing]

# Check missing percentage for these columns and impute if between 10% and 50%
for col in cols_to_impute_unknown:
    missing_percentage = (df_award_notice[col].isnull().sum() / len(df_award_notice)) * 100
    if 10 <= missing_percentage <= 50:
         # Add 'Unknown' as a category if it's a categorical column
        if pd.api.types.isinstance(dtype, pd.CategoricalDtype)(df_award_notice[col]):
            if 'Unknown' not in df_award_notice[col].cat.categories:
                df_award_notice[col] = df_award_notice[col].cat.add_categories('Unknown')
        df_award_notice[col] = df_award_notice[col].fillna('Unknown')
        print(f"Imputed missing values in column '{col}' with 'Unknown'.")

# 5. Address potential outliers in numerical columns (as before, no removal/capping without further investigation)

# Display descriptive statistics after cleaning
display(df_award_notice.describe(include='all'))

No specified columns with high missing values to drop.
Number of columns after dropping: 40


,title,referenceNumber,amendmentNumber,solicitationNumber,contractNumber,publicationDate,contractAwardDate,amendmentDate,contractStartDate,contractEndDate,...,contractingEntityAddressLine,contractingEntityAddressCity,contractingEntityAddressProvince,contractingEntityAddressPostalCode,contractingEntityAddressCountry,endUserEntitiesName,endUserEntitiesAddress,contactInfoName,contactInfoEmail,awardDescription
count,6572,6572,6572.000000,6572,6439,6572,6572,5163,6303,6303,...,6571,6571,6572,6571,6572,6572,6572,6570,6354,6572
unique,4810,6572,NaN,4768,5515,NaN,NaN,NaN,NaN,NaN,...,468,80,25,250,3,123,112,1264,1185,4061
top,E60ZT-180026 ProServices SA - Master Agreement,MX-43089535758,NaN,E60ZT-180026,E60ZT-180026,NaN,NaN,NaN,NaN,NaN,...,"11 Laurier St, Phase III, Place du Portage",Gatineau,Quebec,K1A 0S5,CANADA,Unknown,Unknown,"Fevry, Naomie (SPAC/PSPC) (elle-la / she-her)",naomie.fevry@tpsgc-pwgsc.gc.ca,Unknown
freq,297,1,NaN,298,297,NaN,NaN,NaN,NaN,NaN,...,4038,4596,4817,4385,6570,2294,2294,378,378,1239
mean,NaN,NaN,1.249848,NaN,NaN,2024-10-08 09:20:16.068168192,2024-01-19 10:38:16.287279616,2025-01-23 06:04:48.669378048,2024-01-12 07:21:50.804378880,2026-08-09 14:21:45.663969024,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
min,NaN,NaN,0.000000,NaN,NaN,2024-04-01 00:00:00,2011-12-13 00:00:00,2024-04-01 00:00:00,2012-01-17 00:00:00,2019-01-18 00:00:00,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25%,NaN,NaN,0.000000,NaN,NaN,2024-07-04 00:00:00,2024-04-10 00:00:00,2024-10-31 00:00:00,2024-04-03 00:00:00,2025-03-31 00:00:00,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
50%,NaN,NaN,1.000000,NaN,NaN,2024-10-09 00:00:00,2024-08-14 00:00:00,2025-02-28 00:00:00,2024-08-13 00:00:00,2025-12-04 00:00:00,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
75%,NaN,NaN,1.000000,NaN,NaN,2025-01-17 00:00:00,2024-12-02 00:00:00,2025-04-17 00:00:00,2024-11-29 00:00:00,2027-03-31 00:00:00,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
max,NaN,NaN,26.000000,NaN,NaN,2025-03-31 00:00:00,2027-06-10 00:00:00,2025-07-26 00:00:00,2025-08-01 00:00:00,2100-12-31 00:00:00,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


#### Data Transformation

The difference between contractEndDate and contractStartDate was calculated and stored in the contractDuration column.

The following columns were converted to category datatype for better analysis: contractCurrency, awardStatus, procurementCategory, noticeType, procurementMethod, selectionCriteria, limitedTenderingReason, contractingEntityAddressProvince, contractingEntityAddressCountry and     contactInfoCountry.

The text fields (title and awardDescription) were standardized by converting them to lowercase, removing leading and trailing whitespace, and replacing 'nan' strings with None.


In [50]:
# Calculate contract duration column
df_award_notice['contractDuration'] = df_award_notice['contractEndDate'] - df_award_notice['contractStartDate']

# Convert some columns to categorical datatypes
categorical_cols = ['contractCurrency', 'awardStatus', 'procurementCategory', 'noticeType', 'procurementMethod', 'selectionCriteria', 'limitedTenderingReason', 'contractingEntityAddressProvince', 'contractingEntityAddressCountry', 'contactInfoCountry']

for col in categorical_cols:
    if col in df_award_notice.columns:
        df_award_notice[col] = df_award_notice[col].astype('category')

#Standardize the 'title' and 'awardDescription' columns
cols_to_standardize = ['title', 'awardDescription']

for col in cols_to_standardize:
    if col in df_award_notice.columns:
        df_award_notice[col] = df_award_notice[col].astype(str).str.lower().str.strip().replace('nan', None)

display(df_award_notice.info())
display(df_award_notice.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6572 entries, 0 to 6571
Data columns (total 41 columns):
 #   Column                              Non-Null Count  Dtype          
---  ------                              --------------  -----          
 0   title                               6572 non-null   object         
 1   referenceNumber                     6572 non-null   object         
 2   amendmentNumber                     6572 non-null   int64          
 3   solicitationNumber                  6572 non-null   object         
 4   contractNumber                      6439 non-null   object         
 5   publicationDate                     6572 non-null   datetime64[ns] 
 6   contractAwardDate                   6572 non-null   datetime64[ns] 
 7   amendmentDate                       5163 non-null   datetime64[ns] 
 8   contractStartDate                   6303 non-null   datetime64[ns] 
 9   contractEndDate                     6303 non-null   datetime64[ns] 
 10  contractAmou

None

,title,referenceNumber,amendmentNumber,solicitationNumber,contractNumber,publicationDate,contractAwardDate,amendmentDate,contractStartDate,contractEndDate,...,contractingEntityAddressCity,contractingEntityAddressProvince,contractingEntityAddressPostalCode,contractingEntityAddressCountry,endUserEntitiesName,endUserEntitiesAddress,contactInfoName,contactInfoEmail,awardDescription,contractDuration
0,"pre-qualification for orm consulting, advisory...",MX-43089535758,0,PS23-AG-023,NaN,2024-04-10,2024-04-01,NaT,2024-04-01,2027-03-31,...,Regina,Saskatchewan,S4P 4L3,CANADA,Unknown,Unknown,Adam Gottselig,procurement@fcc.ca,fcc&rsquo;s intention through this rfp is to e...,1094 days
1,"pre-qualification for orm consulting, advisory...",MX-43089535759,0,PS23-AG-023,NaN,2024-04-10,2024-04-01,NaT,2024-04-01,2027-03-31,...,Regina,Saskatchewan,S4P 4L3,CANADA,Unknown,Unknown,Adam Gottselig,procurement@fcc.ca,fcc&rsquo;s intention through this rfp is to e...,1094 days
2,"pre-qualification for orm consulting, advisory...",MX-43089535760,0,PS23-AG-023,NaN,2024-04-10,2024-04-01,NaT,2024-04-01,2027-03-31,...,Regina,Saskatchewan,S4P 4L3,CANADA,Unknown,Unknown,Adam Gottselig,procurement@fcc.ca,fcc&rsquo;s intention through this rfp is to e...,1094 days
3,human centred design,MX-443104980018,0,PS24-AG-001,NaN,2024-04-24,2024-04-24,NaT,2024-04-15,2027-04-14,...,Regina,Saskatchewan,S4P 4L3,CANADA,Unknown,Unknown,Adam Gottselig,procurement@fcc.ca,fcc has an in-house design thinking / human ce...,1094 days
4,rfp – pre-qualification for strategic transfor...,MX-443147876056,0,PS23-AG-018,NaN,2024-06-04,2024-02-05,NaT,2024-02-05,2027-02-04,...,Regina,Saskatchewan,S4P 4L3,CANADA,Unknown,Unknown,Adam Gottselig,procurement@fcc.ca,"effective july 31, 2023, fcc established a tra...",1095 days
